core/io/resource.cpp

```diff
@@ -437,6 +437,98 @@ String Resource::get_id_for_path(const String &p_path) const {
 }
 #endif
 
+bool Resource::_property_can_revert(const StringName &p_name) const {
+	if (!inherits_state.is_valid()) {
+		return false;
+	}
+	bool value_valid;
+	Variant value = get(p_name, &value_valid);
+	if (!value_valid) {
+		return false;
+	}
+	Variant inherits_value = inherits_state->get(p_name, &value_valid);
+	if (!value_valid) {
+		return false;
+	}
+	return bool(Variant::evaluate(Variant::OP_NOT_EQUAL, value, inherits_value));
+}
+
+bool Resource::_property_get_revert(const StringName &p_name, Variant &r_property) const {
+	if (!inherits_state.is_valid()) {
+		return false;
+	}
+
+	bool value_valid;
+	Variant inherits_value = inherits_state->get(p_name, &value_valid);
+	if (value_valid) {
+		r_property = inherits_value;
+	}
+	return value_valid;
+}
+
+void Resource::_validate_property(PropertyInfo &p_property) const {
+	if (p_property.name == "resource_inherits") {
+		p_property.hint_string = get_class();
+	}
+}
+
+bool Resource::setup_inherits_state(const Ref<Resource> &p_resource) {
+	bool ret;
+	if (GDVIRTUAL_CALL(_setup_inherits_state, p_resource, ret)) {
+		return ret;
+	}
+
+	ERR_FAIL_COND_V_MSG(p_resource.is_valid() && !is_class(p_resource->get_class_name()), false, "Resources must be of (or class inherit) the same class when setting up state inheritance");
+
+	if (p_resource.is_valid()) {
+		copy_from(p_resource);
+	}
+	return true;
+}
+
+void Resource::set_inherits_state(const Ref<Resource> &p_resource) {
+	if (setup_inherits_state(p_resource)) {
+		inherits_state = p_resource;
+	}
+}
+
+Ref<Resource> Resource::get_inherits_state() const {
+	return inherits_state;
+}
+
+bool Resource::is_inherited_state_property_value_saved(const StringName &p_name, const Variant &p_value) const {
+	bool ret;
+	if (GDVIRTUAL_CALL(_is_inherited_state_property_value_saved, p_name, p_value, ret)) {
+		return ret;
+	}
+
+	bool default_value_valid;
+	Variant default_value = inherits_state->get(p_name, &default_value_valid);
+	if (!default_value_valid) {
+		return true; // Not a default value, so save.
+	}
+
+	if (default_value.get_type() != Variant::NIL && bool(Variant::evaluate(Variant::OP_EQUAL, p_value, default_value))) {
```

Comment:
What is the reason for skipping `bool Variant::operator==(const Variant&)` here? I don't understand why it's not written as `p_value == default_value`. Can you elaborate?